web/tests/topic_list_data.test.js

```diff
@@ -259,11 +268,11 @@ test("get_list_info unreads", ({override}) => {
     assert.deepEqual(
         list_info.items.map((li) => li.topic_name),
         [
+            "topic 5",
             "topic 0",
             "topic 1",
             "topic 2",
             "topic 3",
-            "topic 5",
```

Comment:
I think the algorithm implementation is probably correct at this point, but what you should do is extend the tests such that the tests would fail with your previous implementation but pass with the new one -- i.e actually exercise the corner cases that the previous version handled incorrectly.